File: Medium/40. Combination Sum II.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] nums, int sum) {
        var res = new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        backtrack(0, sum, nums, new ArrayList<Integer>(), res);
        return res;
    }

    private void backtrack(int start, int left, int[] nums, List comb, List res) {
        if (left == 0) {
            res.add(new ArrayList<>(comb));
            return;
        }
        for (int i = start; i < nums.length; i++) {
            if (nums[i] > left) {
                break;
            }
            if (i > start && nums[i] == nums[i - 1]) {
                continue;
            }
            comb.add(nums[i]);
            backtrack(i + 1, left - nums[i], nums, comb, res);
            comb.removeLast();
        }
    }
}

File: Medium/40. Combination Sum II.kt
class Solution {
    val ans = mutableListOf<List<Int>>()

    fun combinationSum2(candidates: IntArray, target: Int): List<List<Int>> {
        candidates.sort()
        backtrack(candidates, target, 0, mutableListOf())
        return ans
    }

    fun backtrack(
        candidates: IntArray,
        remain: Int,
        start: Int,
        combination: MutableList<Int>
        ) {
            if (remain == 0) {
                ans.add(ArrayList(combination))
                return
            }

            for (i in start until candidates.size) {
                if (i > start && candidates[i] == candidates[i - 1])
                    continue
                if (candidates[i] > remain)
                    break
                
                combination.add(candidates[i])
                backtrack(candidates, remain - candidates[i], i + 1, combination)
                combination.removeAt(combination.size - 1)
            }
        }
}